File: tests/test_libdet.py
from __future__ import annotations

import numpy as np

import libdet


EMPTY = 0
LOWER = 1
UPPER = 2
DOUBLY = 3


def _key(det: np.ndarray) -> bytes:
    return np.ascontiguousarray(det).tobytes()


def _values(dets: np.ndarray, values: np.ndarray) -> dict[bytes, float]:
    return {
        _key(det): float(value)
        for det, value in zip(dets, values, strict=True)
    }


def _as_det(alpha: int, beta: int) -> np.ndarray:
    det = np.zeros((1, 2, 1), dtype=np.uint64)
    det[0, 0, 0] = np.uint64(alpha)
    det[0, 1, 0] = np.uint64(beta)
    return det


def _encode_steps(steps: list[int]) -> np.ndarray:
    alpha = 0
    beta = 0
    for p, step in enumerate(steps):
        if step in (UPPER, DOUBLY):
            alpha |= 1 << p
        if step in (LOWER, DOUBLY):
            beta |= 1 << p
    return _as_det(alpha, beta)[0]


def _guga_paths(norb: int, nelec: int, spin_twice: int) -> np.ndarray:
    steps: list[int] = [EMPTY] * norb
    dets = []

    def rec(p: int, n: int, spin: int) -> None:
        if p == norb:
            if n == nelec and spin == spin_twice:
                dets.append(_encode_steps(steps))
            return

        for step, occ, delta in (
            (EMPTY, 0, 0),
            (LOWER, 1, -1),
            (UPPER, 1, 1),
            (DOUBLY, 2, 0),
        ):
            if n + occ > nelec or spin + delta < 0:
                continue
            steps[p] = step
            rec(p + 1, n + occ, spin + delta)

    rec(0, 0, 0)
    return np.ascontiguousarray(dets, dtype=np.uint64)


def _cg_coeff(
    spin_before: int,
    m_before: int,
    electron_spin: int,
    spin_after: int,
) -> float:
    b = spin_before
    m = m_before
    denom = 2.0 * (b + 1)

    if spin_after == b + 1:
        num = b + m + 2 if electron_spin > 0 else b - m + 2
        return 0.0 if num <= 0 else float(np.sqrt(num / denom))

    if spin_after == b - 1:
        num = b - m if electron_spin > 0 else b + m
        if num <= 0:
            return 0.0
        value = float(np.sqrt(num / denom))
        return -value if electron_spin > 0 else value

    return 0.0


def _decode_steps(det: np.ndarray, norb: int) -> tuple[list[int], list[int]]:
    steps = []
    spin = [0]
    alpha = int(det[0, 0])
    beta = int(det[1, 0])
    for p in range(norb):
        step = ((alpha >> p) & 1) * 2 + ((beta >> p) & 1)
        steps.append(step)
        delta = 1 if step == UPPER else -1 if step == LOWER else 0
        spin.append(spin[-1] + delta)
    return steps, spin


def _canonical_sign(alpha: int, beta: int, norb: int) -> float:
    parity = 0
    for p in range(norb):
        if not ((beta >> p) & 1):
            continue
        for q in range(p + 1, norb):
            parity ^= (alpha >> q) & 1
    return -1.0 if parity else 1.0


def _csf_terms(det: np.ndarray, norb: int) -> dict[tuple[int, int], float]:
    steps, spin = _decode_steps(det, norb)
    current = [(0, 0, 0, 1.0)]

    for p, step in enumerate(steps):
        if step == EMPTY:
            continue
        if step == DOUBLY:
            current = [
                (alpha | (1 << p), beta | (1 << p), m, coeff)
                for alpha, beta, m, coeff in current
            ]
            continue

        next_terms = []
        for alpha, beta, m, coeff in current:
            for sigma in (1, -1):
                c = _cg_coeff(spin[p], m, sigma, spin[p + 1])
                if c == 0.0:
                    continue
                if sigma > 0:
                    next_terms.append(
                        (alpha | (1 << p), beta, m + sigma, coeff * c)
                    )
                else:
                    next_terms.append(
                        (alpha, beta | (1 << p), m + sigma, coeff * c)
                    )
        current = next_terms

    target_m = spin[-1]
    out: dict[tuple[int, int], float] = {}
    for alpha, beta, m, coeff in current:
        if m != target_m or coeff == 0.0:
            continue
        out[(alpha, beta)] = coeff * _canonical_sign(alpha, beta, norb)
    return out


def _guga_ref(
    h1: np.ndarray,
    eri: np.ndarray,
    ecore: float,
    dets: np.ndarray,
) -> np.ndarray:
    norb = h1.shape[0]
    rhf = libdet.Hamiltonian.rhf(h1, eri, ecore=ecore)
    terms = [_csf_terms(det, norb) for det in dets]
    out = np.zeros((len(dets), len(dets)), dtype=np.float64)

    for ibra, bra_terms in enumerate(terms):
        for iket, ket_terms in enumerate(terms):
            value = 0.0
            for (ba, bb), bc in bra_terms.items():
                for (ka, kb), kc in ket_terms.items():
                    value += bc * kc * rhf.hij(_as_det(ba, bb), _as_det(ka, kb))
            out[ibra, iket] = value

    return out


def _pair_index(p: int, q: int) -> int:
    hi, lo = max(p, q), min(p, q)
    return hi * (hi + 1) // 2 + lo


def _small_integrals(norb: int = 3) -> tuple[np.ndarray, np.ndarray, float]:
    rng = np.random.default_rng(31)
    h1 = rng.normal(scale=0.5, size=(norb, norb))
    h1 = 0.5 * (h1 + h1.T)

    npair = norb * (norb + 1) // 2
    pair_eri = rng.normal(scale=0.15, size=(npair, npair))
    pair_eri = 0.5 * (pair_eri + pair_eri.T)

    eri = np.empty((norb, norb, norb, norb), dtype=np.float64)
    for p in range(norb):
        for q in range(norb):
            for r in range(norb):
                for s in range(norb):
                    eri[p, q, r, s] = pair_eri[
                        _pair_index(p, q),
                        _pair_index(r, s),
                    ]

    return h1, eri, 0.17


def test_finite(toy) -> None:
    ham, dets, dense = toy.ham, toy.dets, toy.matrix
    x = np.linspace(-0.8, 1.1, len(dets))
    x2 = np.column_stack((x, x * x))

    np.testing.assert_allclose(ham.matrix(dets).toarray(), dense)
    np.testing.assert_allclose(ham.diag(dets), np.diag(dense))
    np.testing.assert_allclose(ham.matvec(dets, x), dense @ x)
    np.testing.assert_allclose(ham.matvec(dets, x2), dense @ x2)

    empty = dets[:0]
    assert ham.matrix(empty).shape == (0, 0)
    assert ham.matvec(empty, np.empty(0)).shape == (0,)


def test_project(toy) -> None:
    ham, dets, dense, eps = toy.ham, toy.dets, toy.matrix, toy.eps
    kets = dets[:5]
    coeffs = np.array([0.9, -0.7, 0.4, 0.0, 0.2])
    bras = dets[1::3]

    known = ham.project(bras, kets, coeffs)
    np.testing.assert_array_equal(known.bra, bras)
    np.testing.assert_allclose(known.hpsi, dense[1::3, :5] @ coeffs)

    terms = dense[:, :5] * coeffs
    expected = np.where(np.abs(terms) >= eps, terms, 0.0).sum(axis=1)
    external = {_key(det) for det in dets[5:]}
    expected = {
        _key(det): float(value)
        for det, value in zip(dets, expected, strict=True)
        if _key(det) in external and value != 0.0
    }

    expanded = ham.expand(kets, eps, coeffs=coeffs, exclude=kets)
    generated = ham.project(None, kets, coeffs, eps=eps, exclude=kets)

    assert {_key(det) for det in expanded} == set(expected)
    assert _values(generated.bra, generated.hpsi) == expected


def test_conns(toy) -> None:
    ham, dets, dense, eps = toy.ham, toy.dets, toy.matrix, toy.eps
    kets = dets[:6]
    conns = ham.conns(kets, eps)
    cached = ham.conns(kets, eps)
    weight, count = ham.degrees(kets, eps)

    for name in ("x", "ket_ptr", "bra_idx", "h", "weight"):
        np.testing.assert_array_equal(
            np.asarray(getattr(conns, name)),
            np.asarray(getattr(cached, name)),
        )

    pool = np.asarray(conns.x)
    ptr = np.asarray(conns.ket_ptr)
    bra_idx = np.asarray(conns.bra_idx)
    h = np.asarray(conns.h)

    for iket in range(len(kets)):
        expected = {
            _key(dets[ibra]): dense[ibra, iket]
            for ibra in range(len(dets))
            if ibra != iket and abs(dense[ibra, iket]) >= eps
        }
        actual = {
            _key(pool[bra_idx[pos]]): h[pos]
            for pos in range(ptr[iket], ptr[iket + 1])
        }
        assert actual == expected
        assert count[iket] == len(expected)
        np.testing.assert_allclose(weight[iket], sum(map(abs, expected.values())))


def test_sampling(toy) -> None:
    ham, dets, dense = toy.ham, toy.dets, toy.matrix
    kets = dets[:5]
    coeffs = np.array([0.9, -0.7, 0.4, 0.3, -0.2])
    counts = np.array([[5, 4, 3, 2, 1], [1, 2, 3, 4, 5]])

    conns = ham.sample_conns(kets, counts, eps1=np.inf, eps2=0.0, seed=19)
    again = ham.sample_conns(kets, counts, eps1=np.inf, eps2=0.0, seed=19)
    for name in ("x", "ket_ptr", "bra_idx", "h", "count", "weight"):
        np.testing.assert_array_equal(
            np.asarray(getattr(conns, name)),
            np.asarray(getattr(again, name)),
        )

    ptr = np.asarray(conns.ket_ptr)
    hits = np.asarray(conns.count)
    for stream in range(counts.shape[0]):
        for iket in range(len(kets)):
            block = stream * len(kets) + iket
            actual = hits[ptr[block] : ptr[block + 1]].sum()
            expected = counts[stream, iket] if conns.weight[iket] > 0 else 0
            assert actual == expected

    external = np.abs(dense[5:, :5] * coeffs)
    eps1 = float(np.quantile(external[external > 0.0], 0.65))
    sample = ham.sample_project(
        kets,
        coeffs,
        eps1,
        0.0,
        6,
        exclude=kets,
        n_rep=2,
        seed=29,
    )
    repeated = ham.sample_project(
        kets,
        coeffs,
        eps1,
        0.0,
        6,
        exclude=kets,
        n_rep=2,
        seed=29,
    )

    for name in ("rep_ptr", "bra", "hpsi_strong", "hpsi_a", "hpsi_b"):
        np.testing.assert_array_equal(
            np.asarray(getattr(sample, name)),
            np.asarray(getattr(repeated, name)),
        )

    exact = ham.project(sample.bra, kets, coeffs, eps=eps1)
    np.testing.assert_allclose(sample.hpsi_strong, exact.hpsi)
    np.testing.assert_allclose(sample.diag, exact.diag)


def test_guga_singlet() -> None:
    h1 = np.array([[0.3, 0.2], [0.2, 0.7]], dtype=np.float64)
    eri = np.zeros((2, 2, 2, 2), dtype=np.float64)
    ham = libdet.Hamiltonian.guga(
        h1,
        eri,
        n_alpha=1,
        n_beta=1,
    )

    dets = np.zeros((3, 2, 1), dtype=np.uint64)
    dets[0, 0, 0] = np.uint64(1) << np.uint64(0)
    dets[0, 1, 0] = np.uint64(1) << np.uint64(0)
    dets[1, 0, 0] = np.uint64(1) << np.uint64(0)
    dets[1, 1, 0] = np.uint64(1) << np.uint64(1)
    dets[2, 0, 0] = np.uint64(1) << np.uint64(1)
    dets[2, 1, 0] = np.uint64(1) << np.uint64(1)

    expected = np.array(
        [
            [2.0 * h1[0, 0], np.sqrt(2.0) * h1[0, 1], 0.0],
            [
                np.sqrt(2.0) * h1[0, 1],
                h1[0, 0] + h1[1, 1],
                np.sqrt(2.0) * h1[0, 1],
            ],
            [0.0, np.sqrt(2.0) * h1[0, 1], 2.0 * h1[1, 1]],
        ],
        dtype=np.float64,
    )

    matrix = ham.matrix(dets).toarray()
    np.testing.assert_allclose(matrix, expected, atol=1.0e-12)
    np.testing.assert_allclose(ham.diag(dets), np.diag(expected), atol=1.0e-12)
    np.testing.assert_allclose(
        ham.matvec(dets, np.array([0.4, -0.2, 0.8])),
        expected @ np.array([0.4, -0.2, 0.8]),
        atol=1.0e-12,
    )

    coeffs = np.array([0.4, -0.2, 0.8], dtype=np.float64)
    known = ham.project(dets, dets, coeffs)
    np.testing.assert_array_equal(known.bra, dets)
    np.testing.assert_allclose(known.hpsi, expected @ coeffs, atol=1.0e-12)

    expanded = ham.expand(dets[:1], 0.0, exclude=dets[:1])
    assert expanded.shape == (1, 2, 1)
    np.testing.assert_array_equal(expanded[0], dets[1])


def test_guga_conns() -> None:
    h1 = np.array([[0.3, 0.2], [0.2, 0.7]], dtype=np.float64)
    eri = np.zeros((2, 2, 2, 2), dtype=np.float64)
    ham = libdet.Hamiltonian.guga(
        h1,
        eri,
        n_alpha=1,
        n_beta=1,
    )

    kets = np.zeros((1, 2, 1), dtype=np.uint64)
    kets[0, 0, 0] = np.uint64(1) << np.uint64(0)
    kets[0, 1, 0] = np.uint64(1) << np.uint64(0)

    conns = ham.conns(kets, 0.0)
    assert np.asarray(conns.h).shape == (1,)
    np.testing.assert_allclose(conns.h, [np.sqrt(2.0) * h1[0, 1]])

    sample = ham.sample_conns(
        kets,
        np.array([6], dtype=np.int64),
        eps1=np.inf,
        eps2=0.0,
        seed=3,
    )
    assert int(np.asarray(sample.count).sum()) == 6


def test_guga_matrix() -> None:
    h1, eri, ecore = _small_integrals()
    norb = h1.shape[0]
    dets = _guga_paths(norb, nelec=3, spin_twice=1)
    ham = libdet.Hamiltonian.guga(h1, eri, n_alpha=2, n_beta=1, ecore=ecore)

    expected = _guga_ref(h1, eri, ecore, dets)
    actual = ham.matrix(dets).toarray()

    np.testing.assert_allclose(actual, expected, atol=1.0e-11)
    np.testing.assert_allclose(ham.diag(dets), np.diag(expected), atol=1.0e-11)

    x = np.linspace(-0.3, 0.6, len(dets))
    x2 = np.column_stack((x, x[::-1]))
    np.testing.assert_allclose(ham.matvec(dets, x), expected @ x, atol=1.0e-11)
    np.testing.assert_allclose(ham.matvec(dets, x2), expected @ x2, atol=1.0e-11)

    coeffs = np.linspace(0.2, -0.4, len(dets))
    projection = ham.project(dets[::2], dets, coeffs)
    np.testing.assert_array_equal(projection.bra, dets[::2])
    np.testing.assert_allclose(projection.hpsi, expected[::2] @ coeffs, atol=1.0e-11)


def test_guga_external() -> None:
    h1, eri, ecore = _small_integrals()
    norb = h1.shape[0]
    dets = _guga_paths(norb, nelec=3, spin_twice=1)
    ham = libdet.Hamiltonian.guga(h1, eri, n_alpha=2, n_beta=1, ecore=ecore)
    dense = _guga_ref(h1, eri, ecore, dets)

    kets = dets[:4]
    coeffs = np.array([0.6, -0.3, 0.2, 0.5], dtype=np.float64)
    terms = np.abs(dense[:, : len(kets)] * coeffs)
    eps = float(np.quantile(terms[terms > 0.0], 0.45))

    expanded = ham.expand(kets, eps, coeffs=coeffs, exclude=kets)
    generated = ham.project(None, kets, coeffs, eps=eps, exclude=kets)

    expected_hpsi = np.where(
        terms >= eps,
        dense[:, : len(kets)] * coeffs,
        0.0,
    ).sum(axis=1)
    external = {_key(det) for det in dets[len(kets) :]}
    expected = {
        _key(det): float(value)
        for det, value in zip(dets, expected_hpsi, strict=True)
        if _key(det) in external and value != 0.0
    }

    assert {_key(det) for det in expanded} == set(expected)
    assert _values(generated.bra, generated.hpsi) == expected

    conns = ham.conns(kets, eps)
    weight, count = ham.degrees(kets, eps)
    pool = np.asarray(conns.x)
    ptr = np.asarray(conns.ket_ptr)
    bra_idx = np.asarray(conns.bra_idx)
    h = np.asarray(conns.h)

    for iket in range(len(kets)):
        expected_conn = {
            _key(dets[ibra]): dense[ibra, iket]
            for ibra in range(len(dets))
            if ibra != iket and abs(dense[ibra, iket]) >= eps
        }
        actual_conn = {
            _key(pool[bra_idx[pos]]): h[pos]
            for pos in range(ptr[iket], ptr[iket + 1])
        }
        assert actual_conn == expected_conn
        assert count[iket] == len(expected_conn)
        np.testing.assert_allclose(weight[iket], sum(map(abs, expected_conn.values())))

    sample = ham.sample_conns(
        kets,
        np.array([4, 3, 2, 1], dtype=np.int64),
        eps1=np.inf,
        eps2=0.0,
        seed=41,
    )
    hits = np.asarray(sample.count)
    ptr = np.asarray(sample.ket_ptr)
    for iket in range(len(kets)):
        actual = hits[ptr[iket] : ptr[iket + 1]].sum()
        expected_count = [4, 3, 2, 1][iket] if sample.weight[iket] > 0 else 0
        assert actual == expected_count

File: tests/cpp/test_libdet.cpp
#include <libdet/hamiltonian.hpp>

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <limits>
#include <stdexcept>
#include <vector>

using libdet::DetBatchView;
using libdet::DetRef;
using libdet::Hamiltonian;
using libdet::i64;
using libdet::u64;

void near(double actual, double expected, const char* message) {
    const double tol = 4.0e-12 * (1.0 + std::abs(expected));
    if (std::abs(actual - expected) > tol) {
        throw std::runtime_error(message);
    }
}

std::size_t pair_id(int p, int q) {
    const int hi = std::max(p, q);
    const int lo = std::min(p, q);
    return static_cast<std::size_t>(hi * (hi + 1) / 2 + lo);
}

std::size_t eri_id(int p, int q, int r, int s) {
    return pair_id(
        static_cast<int>(pair_id(p, q)),
        static_cast<int>(pair_id(r, s))
    );
}

std::vector<double> make_h1(int norb) {
    std::vector<double> out(static_cast<std::size_t>(norb * norb), 0.0);
    for (int p = 0; p < norb; ++p) {
        for (int q = 0; q <= p; ++q) {
            const double h =
                0.11 + 0.07 * (p + 1) - 0.03 * (q + 2)
                + 0.015 * ((p + 2 * q) % 5);
            out[static_cast<std::size_t>(p * norb + q)] = h;
            out[static_cast<std::size_t>(q * norb + p)] = h;
        }
    }
    return out;
}

std::vector<double> make_eri(int norb) {
    const int npair = norb * (norb + 1) / 2;
    std::vector<double> out(
        static_cast<std::size_t>(npair * (npair + 1) / 2),
        0.0
    );
    for (int p = 0; p < norb; ++p) {
        for (int q = 0; q < norb; ++q) {
            for (int r = 0; r < norb; ++r) {
                for (int s = 0; s < norb; ++s) {
                    out[eri_id(p, q, r, s)] =
                        0.012
                        * (1 + ((3 * p + 5 * q + 7 * r + 11 * s) % 23));
                }
            }
        }
    }
    return out;
}

std::vector<u64> make_basis(int norb) {
    std::vector<u64> out;
    for (int i = 0; i < norb; ++i) {
        for (int j = i + 1; j < norb; ++j) {
            const u64 alpha = (u64{1} << i) | (u64{1} << j);
            for (int k = 0; k < norb; ++k) {
                out.push_back(alpha);
                out.push_back(u64{1} << k);
            }
        }
    }
    return out;
}

DetBatchView view(const std::vector<u64>& words) {
    return {words.data(), words.size() / 2u, 1u};
}

int find_state(DetBatchView states, DetRef target) {
    for (std::size_t i = 0; i < states.n_dets; ++i) {
        const DetRef state = states[i];
        if (
            state.alpha()[0] == target.alpha()[0]
            && state.beta()[0] == target.beta()[0]
        ) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

std::vector<double> dense(
    const Hamiltonian& ham,
    DetBatchView basis
) {
    const std::size_t n = basis.n_dets;
    std::vector<double> out(n * n, 0.0);
    for (std::size_t bra = 0; bra < n; ++bra) {
        for (std::size_t ket = 0; ket < n; ++ket) {
            out[bra * n + ket] = ham.hij(basis[bra], basis[ket]);
        }
    }
    return out;
}

double sum_degree(
    const std::vector<double>& matrix,
    std::size_t n,
    std::size_t ket,
    double eps1,
    double eps2 = std::numeric_limits<double>::infinity()
) {
    double out = 0.0;
    for (std::size_t bra = 0; bra < n; ++bra) {
        if (bra == ket) continue;
        const double h = std::abs(matrix[bra * n + ket]);
        if (h >= eps1 && h < eps2) out += h;
    }
    return out;
}

void check_matrix(
    const Hamiltonian& ham,
    DetBatchView basis,
    const std::vector<double>& expected
) {
    const std::size_t n = basis.n_dets;
    const auto diag = ham.diag(basis);
    const auto matrix = ham.matrix(basis, basis);
    std::vector<double> vec(n * 2u, 0.0);
    for (std::size_t i = 0; i < vec.size(); ++i) {
        vec[i] = 0.1 * static_cast<double>(i + 1u);
    }
    const auto product = ham.matvec(basis, basis, {vec.data(), n});
    const auto products = ham.matmat(basis, basis, vec, 2u);

    for (std::size_t bra = 0; bra < n; ++bra) {
        near(diag[bra], expected[bra * n + bra], "diag");

        std::vector<double> row(n, 0.0);
        for (i64 p = matrix.indptr[bra]; p < matrix.indptr[bra + 1u]; ++p) {
            const std::size_t pos = static_cast<std::size_t>(p);
            row[static_cast<std::size_t>(matrix.indices[pos])] = matrix.data[pos];
        }

        double value = 0.0;
        double value0 = 0.0;
        double value1 = 0.0;
        for (std::size_t ket = 0; ket < n; ++ket) {
            const double h = expected[bra * n + ket];
            near(row[ket], h, "matrix");
            value += h * vec[ket];
            value0 += h * vec[2u * ket];
            value1 += h * vec[2u * ket + 1u];
        }
        near(product[bra], value, "matvec");
        near(products[2u * bra], value0, "matmat");
        near(products[2u * bra + 1u], value1, "matmat");
    }
}

void check_conn(
    const Hamiltonian& ham,
    DetBatchView basis,
    const std::vector<double>& matrix,
    double eps
) {
    const std::size_t n = basis.n_dets;
    const auto conn = ham.conn(basis, eps);
    const auto bras = view(conn.bra);

    for (std::size_t ket = 0; ket < n; ++ket) {
        near(conn.diag[ket], matrix[ket * n + ket], "conn diag");
        near(conn.degree[ket], sum_degree(matrix, n, ket, eps), "degree");

        double previous = std::numeric_limits<double>::infinity();
        std::vector<double> column(n, 0.0);
        for (i64 p = conn.ptr[ket]; p < conn.ptr[ket + 1u]; ++p) {
            const std::size_t pos = static_cast<std::size_t>(p);
            const int bra = find_state(basis, bras[n + pos]);
            if (bra < 0 || std::abs(conn.h[pos]) > previous) {
                throw std::runtime_error("conn order");
            }
            previous = std::abs(conn.h[pos]);
            column[static_cast<std::size_t>(bra)] = conn.h[pos];
        }

        for (std::size_t bra = 0; bra < n; ++bra) {
            const double h = matrix[bra * n + ket];
            const double value = bra != ket && std::abs(h) >= eps ? h : 0.0;
            near(column[bra], value, "conn value");
        }
    }
}

void check_project(
    const Hamiltonian& ham,
    DetBatchView basis,
    const std::vector<double>& matrix
) {
    const std::size_t n = basis.n_dets;
    const DetBatchView kets{basis.data, 4u, basis.nword};
    const std::vector<double> scale{0.4, -0.2, 0.7, 0.3};
    const auto fixed = ham.project(basis, kets, scale, 0.0);

    for (std::size_t bra = 0; bra < n; ++bra) {
        double value = 0.0;
        for (std::size_t ket = 0; ket < kets.n_dets; ++ket) {
            value += matrix[bra * n + ket] * scale[ket];
        }
        near(fixed.hpsi[bra], value, "project");
    }

    const double eps = 0.08;
    const auto words = ham.expand(kets, eps, scale, &kets);
    const auto expanded = view(words);
    std::vector<bool> expected(n, false);
    for (std::size_t bra = kets.n_dets; bra < n; ++bra) {
        for (std::size_t ket = 0; ket < kets.n_dets; ++ket) {
            if (std::abs(matrix[bra * n + ket] * scale[ket]) >= eps) {
                expected[bra] = true;
            }
        }
    }

    std::size_t count = 0;
    for (bool value : expected) count += value ? 1u : 0u;
    if (expanded.n_dets != count) throw std::runtime_error("expand size");
    for (std::size_t i = 0; i < expanded.n_dets; ++i) {
        const int bra = find_state(basis, expanded[i]);
        if (bra < 0 || !expected[static_cast<std::size_t>(bra)]) {
            throw std::runtime_error("expand value");
        }
    }
}

void check_sample(
    const Hamiltonian& ham,
    DetBatchView basis,
    const std::vector<double>& matrix,
    double eps1,
    double eps2
) {
    const std::size_t n = basis.n_dets;
    const std::vector<i64> counts(2u * n, 3);
    const auto sample_a = ham.sample_conn(
        basis,
        counts,
        2u,
        eps1,
        eps2,
        11
    );
    const auto sample_b = ham.sample_conn(
        basis,
        counts,
        2u,
        eps1,
        eps2,
        11
    );
    if (
        sample_a.bra != sample_b.bra
        || sample_a.ptr != sample_b.ptr
        || sample_a.h != sample_b.h
        || sample_a.degree != sample_b.degree
    ) {
        throw std::runtime_error("sample seed");
    }

    const auto sample_bra = view(sample_a.bra);
    for (std::size_t row = 0; row < 2u * n; ++row) {
        const std::size_t ket = row % n;
        near(
            sample_a.degree[ket],
            sum_degree(matrix, n, ket, eps2, eps1),
            "sample degree"
        );
        for (i64 p = sample_a.ptr[row]; p < sample_a.ptr[row + 1u]; ++p) {
            const std::size_t pos = static_cast<std::size_t>(p);
            const int bra = find_state(basis, sample_bra[n + pos]);
            if (bra < 0) throw std::runtime_error("sample bra");
            const double h = matrix[static_cast<std::size_t>(bra) * n + ket];
            if (!(std::abs(h) >= eps2 && std::abs(h) < eps1)) {
                throw std::runtime_error("sample window");
            }
            near(sample_a.h[pos], h, "sample value");
        }
    }

    const std::vector<i64> draws(n, 3);
    const auto local_a = ham.local_conn(basis, eps1, eps2, draws, 17);
    const auto local_b = ham.local_conn(basis, eps1, eps2, draws, 17);
    if (
        local_a.bra != local_b.bra
        || local_a.strong_ptr != local_b.strong_ptr
        || local_a.strong_h != local_b.strong_h
        || local_a.weak_ptr != local_b.weak_ptr
        || local_a.weak_coeff != local_b.weak_coeff
    ) {
        throw std::runtime_error("local seed");
    }

    const auto local_bra = view(local_a.bra);
    const std::size_t strong_n =
        static_cast<std::size_t>(local_a.strong_ptr.back());
    for (std::size_t ket = 0; ket < n; ++ket) {
        near(
            local_a.strong_degree[ket],
            sum_degree(matrix, n, ket, eps1),
            "strong degree"
        );
        for (
            i64 p = local_a.strong_ptr[ket];
            p < local_a.strong_ptr[ket + 1u];
            ++p
        ) {
            const std::size_t pos = static_cast<std::size_t>(p);
            const int bra = find_state(basis, local_bra[n + pos]);
            if (bra < 0) throw std::runtime_error("strong bra");
            const double h = matrix[static_cast<std::size_t>(bra) * n + ket];
            if (std::abs(h) < eps1) throw std::runtime_error("strong window");
            near(local_a.strong_h[pos], h, "strong value");
        }
        for (
            i64 p = local_a.weak_ptr[ket];
            p < local_a.weak_ptr[ket + 1u];
            ++p
        ) {
            const std::size_t pos = static_cast<std::size_t>(p);
            const int bra = find_state(
                basis,
                local_bra[n + strong_n + pos]
            );
            if (bra < 0) throw std::runtime_error("weak bra");
            const double h = matrix[static_cast<std::size_t>(bra) * n + ket];
            if (
                !(std::abs(h) >= eps2 && std::abs(h) < eps1)
                || local_a.weak_coeff[pos] * h <= 0.0
            ) {
                throw std::runtime_error("weak window");
            }
        }
    }
}

int main() {
    const int norb = 5;
    const auto h1 = make_h1(norb);
    const auto eri = make_eri(norb);
    const auto words = make_basis(norb);
    const auto basis = view(words);
    const Hamiltonian ham(h1, norb, eri, -0.13);
    const auto matrix = dense(ham, basis);

    check_matrix(ham, basis, matrix);
    check_conn(ham, basis, matrix, 0.10);
    check_project(ham, basis, matrix);
    check_sample(ham, basis, matrix, 0.10, 0.025);
    return 0;
}
